java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -636,8 +646,8 @@ private <T extends IncrementalTask> void createTasks(Datastore datastore,
         ShardRetryState<T> retryState = ShardRetryState.createFor(taskState);
         tx.put(IncrementalTaskState.Serializer.toEntity(tx, taskState),
           ShardRetryState.Serializer.toEntity(tx, retryState));
-        scheduleWorkerTask(settings, taskState, null);
         tx.commit();
+        scheduleWorkerTask(settings, taskState, null);
```

Comment:
this should help too; don't enqueue task until AFTER we've successfully committed. 

We do have a risk that if enqueuing task fails, shard will just stall out, right? Nothing to recover it and ensure it eventually gets executed.